lib/sidekiq/api.rb

```diff
@@ -520,35 +526,41 @@ def find_job(jid)
     end
 
     def delete(score, jid = nil)
-      if jid
-        elements = Sidekiq.redis do |conn|
-          conn.zrangebyscore(name, score, score)
-        end
+      jid ? delete_by_jid(score, jid) : delete_by_score(score, score)
+    end
 
-        elements_with_jid = elements.map do |element|
-          message = Sidekiq.load_json(element)
+    def delete_by_value(name, value)
+      Sidekiq.redis do |conn|
+        ret = conn.zrem(name, value)
+        @_size -= 1 if ret
+        ret
+      end
+    end
 
-          if message["jid"] == jid
-            _, @_size = Sidekiq.redis do |conn|
-              conn.multi do
-                conn.zrem(name, element)
-                conn.zcard name
-              end
-            end
-          end
-        end
-        elements_with_jid.count != 0
-      else
-        count, @_size = Sidekiq.redis do |conn|
-          conn.multi do
-            conn.zremrangebyscore(name, score, score)
-            conn.zcard name
+    def delete_by_jid(score, jid)
+      elements = Sidekiq.redis do |conn|
+        conn.zrangebyscore(name, score, score)
+      end
+      elements.each do |element|
+        message = Sidekiq.load_json(element)
+        if message["jid"] == jid
+          break Sidekiq.redis do |conn|
+            ret = conn.zrem(name, element)
+            @_size -= 1 if ret
+            ret
           end
         end
-        count != 0
+        false
       end
     end
 
+    def delete_by_score(score_min, score_max)
```

Comment:
@pik I agree with @jonhyman, this is a dangerous option and should be removed even if the current API allows it.  Maybe it's a one in a million chance but at scale it will happen and those are the data loss issues that are so painful to track down.